runtime/ops/otel.rs

```diff
@@ -0,0 +1,582 @@
+      tokio::sync::mpsc::error::TrySendError::Full(_) => {
+        opentelemetry_sdk::runtime::TrySendError::ChannelFull
+      }
+      tokio::sync::mpsc::error::TrySendError::Closed(_) => {
+        opentelemetry_sdk::runtime::TrySendError::ChannelClosed
+      }
+    })
+  }
+}
+
+pub struct BatchMessageChannelReceiver<T> {
+  receiver: tokio::sync::mpsc::Receiver<T>,
+}
+
+impl<T> From<tokio::sync::mpsc::Receiver<T>>
+  for BatchMessageChannelReceiver<T>
+{
+  fn from(receiver: tokio::sync::mpsc::Receiver<T>) -> Self {
+    Self { receiver }
+  }
+}
+
+impl<T> Stream for BatchMessageChannelReceiver<T> {
+  type Item = T;
+
+  fn poll_next(
+    mut self: Pin<&mut Self>,
+    cx: &mut Context<'_>,
+  ) -> Poll<Option<Self::Item>> {
+    self.receiver.poll_recv(cx)
+  }
+}
+
+fn otel_create_globals(
+  config: OtelConfig,
+  op_state: &mut OpState,
+) -> anyhow::Result<()> {
+  // Parse the `OTEL_EXPORTER_OTLP_PROTOCOL` variable. The opentelemetry_*
+  // crates don't do this automatically.
+  // TODO(piscisaureus): enable GRPC support.
+  let protocol = match env::var("OTEL_EXPORTER_OTLP_PROTOCOL").as_deref() {
+    Ok("http/protobuf") => Protocol::HttpBinary,
+    Ok("http/json") => Protocol::HttpJson,
+    Ok("") | Err(env::VarError::NotPresent) => {
+      return Ok(());
+    }
+    Ok(protocol) => {
+      return Err(anyhow!(
+        "Env var OTEL_EXPORTER_OTLP_PROTOCOL specifies an unsupported protocol: {}",
+        protocol
+      ));
+    }
+    Err(err) => {
+      return Err(anyhow!(
+        "Failed to read env var OTEL_EXPORTER_OTLP_PROTOCOL: {}",
+        err
+      ))
+    }
+  };
+
+  // Define the resource attributes that will be attached to all log records.
+  // These attributes are sourced as follows (in order of precedence):
+  //   * The `service.name` attribute from the `OTEL_SERVICE_NAME` env var.
+  //   * Additional attributes from the `OTEL_RESOURCE_ATTRIBUTES` env var.
+  //   * Default attribute values defined here.
+  // TODO(piscisaureus): add more default attributes (e.g. script path).
+  let mut resource = Resource::default();
+  // The default service name assigned by `Resource::default()`, if not
+  // otherwise specified via environment variables, is "unknown_service".
+  // Override this with the current crate name and version.
+  if resource
+    .get(Key::from_static_str(SERVICE_NAME))
+    .filter(|service_name| service_name.as_str() != "unknown_service")
+    .is_none()
+  {
+    resource = resource.merge(&Resource::new(
+      [
+        (SERVICE_NAME, config.default_service_name),
+        (SERVICE_VERSION, config.default_service_version),
+      ]
+      .into_iter()
+      .map(|(k, v)| KeyValue::new(k, v)),
+    ))
+  }
+
+  // The OTLP endpoint is automatically picked up from the
+  // `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable. Additional headers can
+  // be specified using `OTEL_EXPORTER_OTLP_HEADERS`.
+
+  let client = reqwest::Client::new();
```

Comment:
Could we try to use hyper directly? We made some efforts this year to move away from using `reqwest` in favor of `hyper` directly to have greater control. Maybe it's fine for the first pass, but if so I'd like to have an issue for a future TODO for that.

Reply:
Should be a simple change: https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-http/src/lib.rs#L104